runtime/vm/UpcallThunkMem.cpp

```diff
@@ -265,7 +266,54 @@ allocateThunkHeap(J9UpcallMetaData *data)
 }
 
 /**
- * Compute the hash code (namely the thunk address value) for the supplied J9UpcallMetaDataEntry
+ * @brief Release the thunk heap list if exists.
+ *
+ * @param vm a pointer to J9JavaVM
+ * @return void
+ *
+ * Note:
+ * This function empty the thunk heap list by cleaning up all resources
+ * created via allocateUpcallThunkMemory, including the generated thunk
+ * and the corresponding metadata of each entry in the hashtable.
+ */
+extern "C" void
+releaseThunkHeap(J9JavaVM * vm)
+{
+	J9UpcallThunkHeapList *thunkHeapHead = vm->thunkHeapHead;
+	J9UpcallThunkHeapList *thunkHeapNode = NULL;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	UDATA byteAmount = j9vmem_supported_page_sizes()[0];
+
+	if (NULL != thunkHeapHead->metaDataHashTable) {
+		J9HashTable *metaDataHashTable = thunkHeapHead->metaDataHashTable;
+		UDATA entryCount = hashTableGetCount(metaDataHashTable);
```

Comment:
> Won't hashTableForEachDo() simply do nothing for an empty hash table?
In theory that's correct but the code in `hashTableForEachDo()` still relies on the count to determine whether to move forward to check the hashtable which is still a bunch of code/assignment before returning null for a empty hashtable

See the related code 
at https://github.com/eclipse-openj9/openj9-omr/blob/9083c8237ac215927ac55b5db256780132983136/util/hashtable/hashtable.c#L829
```
void
hashTableForEachDo(J9HashTable *table, J9HashTableDoFn doFn, void *opaque)
{
...
	node = hashTableStartDo(table,  &walkState); <------
...
```
and https://github.com/eclipse-openj9/openj9-omr/blob/9083c8237ac215927ac55b5db256780132983136/util/hashtable/hashtable.c#L1005
```
...
void *
hashTableStartDo(J9HashTable *table,  J9HashTableState *handle)
{
	void *result = NULL;
	uint32_t numberOfListNodes = table->numberOfNodes - table->numberOfTreeNodes; <-----
	HASHTABLE_DEBUG_PORT(table->portLibrary);

	memset(handle, 0, sizeof(J9HashTableState));
	handle->table = table;
	handle->bucketIndex = 0;
	handle->pointerToCurrentNode = &table->nodes[0];
	handle->didDeleteCurrentNode = FALSE;
	handle->iterateState = J9HASH_TABLE_ITERATE_STATE_LIST_NODES;
...
	} else {
		/* It would be convenient if we could iterate and perform doRemoves on avl trees, but currently that API is not supported, so we just
		 * have to iterate the treeNode pool
		 */
		if (numberOfListNodes > 0) {
...
		} else if (table->numberOfTreeNodes > 0) {
...
		} else {
			result = NULL;
			handle->iterateState = J9HASH_TABLE_ITERATE_STATE_FINISHED;
		}
...
return result; <----- return null for empty hashtable
}
```
So hashTableGetCount() at 
https://github.com/eclipse-openj9/openj9-omr/blob/9083c8237ac215927ac55b5db256780132983136/util/hashtable/hashtable.c#L1205 is more straightforward as compared to `hashTableForEachDo`.
```
uint32_t
hashTableGetCount(J9HashTable *table)
{
	return table->numberOfNodes;
}
```

Reply:
If we prefer `hashTableForEachDo`, the better way is to move the count check to `hashTableForEachDo` before calling `hashTableStartDo` in OMR. e.g.
```
hashTableForEachDo(J9HashTable *table, J9HashTableDoFn doFn, void *opaque)
{
...
      if (table->numberOfNodes > 0) { <---------------------
	node = hashTableStartDo(table,  &walkState);
      }
while (NULL != node) {
		if (doFn(node, opaque)) {
			hashTableDoRemove(&walkState);
		}
		node = hashTableNextDo(&walkState);
	}
```

Reply:
I think checking `numberOfNodes` early would be an improvement to `hashTableStartDo()` and `hashTableForEachDo()`, and possibly other hash table functions, but I agree that, and the subsequent updates to client code should be done separately from this change.

Reply:
Removed `hashTableGetCount()` related code and created an issue at https://github.com/eclipse-openj9/openj9-omr/issues/211 to keep track of the hashtable related improvement.